### PR TITLE
Update and test `EnumerableSet`

### PR DESCRIPTION
**Pull request by 0xCourtney**

### CODE REVIEW

File: test/data/EnumerableSet.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+          await instance['add(bytes32)'](twoBytes32);
+
+          expect(await instance['at(uint256)'](0)).to.equal(zeroBytes32);
+          expect(await instance['at(uint256)'](1)).to.equal(oneBytes32);
+          expect(await instance['at(uint256)'](2)).to.equal(twoBytes32);
+        });
+
+        describe('reverts if', function () {
+          it('index out of bounds', async () => {
+            await expect(instance['at(uint256)'](0)).to.be.revertedWith(
```
Comment (NouDaimon) [inline test/data/EnumerableSet.ts:38]: Perhaps wait until #139 is merged before continuing with tests?

Comment (NouDaimon) [reply test/data/EnumerableSet.ts:38]: Actually, this is minor and can be retroactively fixed.

File: test/data/EnumerableSet.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+          await instance['remove(bytes32)'](zeroBytes32);
+          expect(await instance['length()']()).to.be.equal(0);
+          expect(await instance['indexOf(bytes32)'](zeroBytes32)).to.be.equal(
+            ethers.constants.MaxUint256,
+          );
+        });
+
+        describe('#toArray()', () => {
+          it('returns the set as an array', async () => {
+            await instance['add(bytes32)'](zeroBytes32);
```
[the remark this answers is not in the document]
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:181]: Are you talking about using a different matcher?  Can't remember which one - maybe `to.deep.have.members` or something.

File: test/data/EnumerableSet.ts (hunk 1 above)
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:38]: Merged in that branch, can be fixed now.

File: test/data/EnumerableSet.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:181]: I agree with that in theory, but it only really matters when writing spec.

File: test/data/EnumerableSet.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    let instance: EnumerableSetBytes32Mock;
+    let deployer: SignerWithAddress;
+
+    beforeEach(async () => {
+      [deployer] = await ethers.getSigners();
+      instance = await new EnumerableSetBytes32Mock__factory(deployer).deploy();
+    });
+
+    describe('__internal', () => {
+      const zeroBytes32 = bnToBytes32(ethers.constants.Zero);
```
Comment (ItsNickBarry) [inline test/data/EnumerableSet.ts:21]: There's also `ethers.constants.HashZero`, but your way is probably more clear.

File: test/data/EnumerableSet.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:181]: Never mind, you're talking about adding the elements in a different order and testing that they're output in sorted order, correct?  In that case, yes, mix up the insertion order and add a comment explaining why.

[the remark this answers is not in the document]
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:181]: Do the same for the `at` test.

[the remark this answers is not in the document]
Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:181]: Deleted some previous comments - forgot which PR I was reviewing.

File: test/data/EnumerableSet.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.true;
+        });
+
+        it('returns false if value is not removed', async () => {
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.false;
+        });
+
+        it('removes value from index mapping and array', async () => {
```
Comment (ItsNickBarry) [inline test/data/EnumerableSet.ts:155]: I would probably move these tests to the `length` and `indexOf` blocks.  Seems like too many things are happening here.  I don't feel strongly about it, however.

**Review by ItsNickBarry — COMMENTED** [review]
Tests seem good, okay to make copies for `AddressSet` and `UintSet`.

File: test/data/EnumerableSet.ts (hunk 4 above)
Comment (NouDaimon) [reply test/data/EnumerableSet.ts:155]: I think the purpose is to test all effects of the function in an effort for self-containment.

The remove function explicitly alters these properties.

Comment (ItsNickBarry) [reply test/data/EnumerableSet.ts:155]: If we separate test concerns perfectly, the `remove` tests must test the side effects to `length` and `indexOf`, and both the `length` and `indexOf` tests must test what happens when `remove` is called.  This is a bit too redundant, so we must choose which approach makes more sense.  There's no right answer.

Comment (0xCourtney) [reply test/data/EnumerableSet.ts]: I removed the "removes value from index mapping and array" test case. The array length is already checked when values are added/ removed from the set in the "returns length of enumerable set" test case. I modified the test case "returns index of the value" to check that the mapping is updated when values are added/ removed from the set.

https://github.com/solidstate-network/solidstate-solidity/pull/144/commits/ffb9354f2cb3b9c6855cf7689266180c7c95d3c8

File: test/data/EnumerableSet.ts (hunk 3 above)
Comment (0xCourtney) [reply test/data/EnumerableSet.ts:26]: Awesome, I left it as is for consistency but I will be sure to keep this in mind when I need a byte32 sized zero hash.

File: test/data/EnumerableSet.ts (hunk 4 above)
Comment (0xCourtney) [reply test/data/EnumerableSet.ts]: I will also update this test case in the BinaryHeap PR, as its checked there as well.

**Review by NouDaimon — APPROVED** [review]

**Review by ItsNickBarry — APPROVED** [review]
Okay to merge, but not until after #143.

File: contracts/data/EnumerableSet.sol [hunk 5]
```diff
@@ -150,6 +150,57 @@ library EnumerableSet {
...
+    function toArray(UintSet storage set)
+        internal
+        view
+        returns (uint256[] memory)
+    {
+        uint256 len = _length(set._inner);
+        uint256[] memory arr = new uint256[](len);
+
+        unchecked {
+            for (uint256 index = 0; index < len; index++) {
```
Comment (NouDaimon) [inline contracts/data/EnumerableSet.sol:196]: Should look like:
```
for (uint256 index; index < len;) {
      arr[index] = at(set, index);
      ++i;
}
```

No need to initialize the index.

Comment (NouDaimon) [reply contracts/data/EnumerableSet.sol:196]: Same applies for other for loops.

Comment (0xCourtney) [reply contracts/data/EnumerableSet.sol]: https://github.com/solidstate-network/solidstate-solidity/pull/144/commits/f83e3986fb7d6d767a7130e61e0eef3842296c15

**Review by NouDaimon — APPROVED** [review]

**Pull request merged**